### ISSUE TITLE
Job cache on the master contains only partial job result information

### ISSUE THREAD

**Issue by szjur** [saltstack/salt#35691]
### Description of Issue/Question

I start a job to execute cmd.run with arg "rm /nonexistent" on a minion.
I get it's result on the eventbus:

{'tag': 'salt/job/20160823124251202948/ret/minionname.my.domain',
'data': {'fun_args': None, 'jid': '20160823124251202948', 'return': "rm:
cannot remove `/nonexistent': No such file or directory", 'retcode': 1,
'success': True, '_stamp': '2016-08-23T10:42:51.473637', 'fun':
'cmd.run', 'master_id': 'mastername.my.domain', 'id':
'minionname.my.domain'}}

So I get the command output ('return') and also 'retcode' plus 'success'.

However, when I look in the job cache files on the master in
/var/cache/salt/master/jobs, the return.p file for that minion only
contains the 'return' part of the result. There is no way that any of
the runner modules such as jobs.list_job, jobs.lookup_jid or
jobs.print_job  returned the missing  information.

Amazingly, when I turn on the local job cache on the minion, then it
stores the complete information in

```
/var/cache/salt/minion/minion_jobs/<jobid>/return.p
```

Consequently, saltutil.find_cached_job executed on that  minion returns
complete information about that job.

The question is, was that job cache on the master intentionally made
this way that it contains only partial information and  if  so, why is that?

I would imagine that the job cache could be used to retrieve the job
result in the same form as it  originally came in. In my use case that
result of cmd.run without the retcode is of very limited usability.
### Setup

Any setup with a master and a minion should do. Default job cache on the master.
### Steps to Reproduce Issue

Run
   salt -L minion.name cmd.run "rm /nonexistent"

Then try to get that job result including the retcode from the job cache on the master.
### Versions Report

2016.3.2

**Comment by DmitryKuzmenko** [comment]
Job local cache structure is the following for each job:
```
|- jid (contains job id)
|- .load.p (serialized load)
|- .minions.p (minions list)
|- <minion_id>
|   |- out.p (job output)
|   `- return.p (job return)
`- <minion_id>
...
```

The file `return.p` contains the job return that is an error string in the described case. In Syndic ret I've changed the return field in loads to keep not the only `return` value but a dict of return-related values like `{'return': ..., 'success': ..., 'retcode': ...}.
I can make the same here. Also it will be needed to update all the `returners` modules.

@cachedout, @thatch45 what do you think about it?

**Comment by thatch45** [comment]
I think it is a good idea, we should always push all of the return metadata

**Comment by tsaridas** [comment]
@DmitryKuzmenko I see that the results are shown in the load but not saved. Are you working on this ? Is there any ETA ?
thnx

**Comment by cachedout** [comment]
Closed via #40697
